api/client/websocketupgradeproto/protocol.go

```diff
@@ -0,0 +1,344 @@
+/*
+Copyright 2025 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package websocketupgradeproto
+
+import (
+	"context"
+	"errors"
+	"io"
+	"log/slog"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gobwas/ws"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/constants"
+)
+
+const (
+	ComponentTeleport = "teleport"
+	ComponentClient   = "client"
+)
+
+// connectionType represents the side of the connection.
+// It can be either a server or a client connection.
+// This is used to determine how to handle WebSocket frames, especially
+// for the ping/pong frames.
+type connectionType int
```

Comment:
nit: maybe use https://pkg.go.dev/github.com/gobwas/ws#State ?

Reply:
ws.State has other states like fragmented and extended that I do not want to use or allow anyone setting them.

Reply:
sure. if we gonna use some other helpers from the ws lib say `wsutils.HandleControlMessage`, it would be easier to use `ws.State`. but we are not anyway